lib/1.4/utility.dml

```diff
@@ -192,15 +192,105 @@ template _init_val_power_on_reset is (init_val, power_on_reset) {
     }
 }
 
+/**
+### signal_port
+
+<section>
+
+#### Description
+
+Implements a signal interface with saved state. The current state of the signal
+is stored in the saved boolean `high` and an error is logged if the signal is
+raised or lowered when already high or low. The methods `signal_raise` and
+`signal_lower` can be overridden to add additional side effects.
+
+#### Related Templates
+
+poreset, hreset, sreset
+
+</section>
+*/
+template signal_port {
+    implement signal {
+        saved bool high;
+        method signal_raise() default {
+            if (high)
+                log error: "%s raised when already high", qname;
+            else
+                high = true;
+        }
+        method signal_lower() default {
+            if (!high)
+                log error: "%s lowered when already low", qname;
+            else
+                high = false;
+        }
+    }
+}
+
+/**
+### signal_connect
+
+<section>
+
+#### Description
+
+Implements a connect with a signal interface, with saved state. The current
+state of the signal is stored in the saved boolean `signal.high`. If the
+connect is changed while `signal.high` is `true`, the `signal_raise` method
+will be called on the new object. This behaviour can be changed by overriding
+the `set` method. The template defines the following methods, which are default
+and can be overridden:
+
+* `raise()`: Calls `signal_raise` if `signal.high` is false, and sets
+   `signal.high` to `true`.
+
+* `lower()`: Calls `signal_lower` if `signal.high` is true, and sets
+   `signal.high` to `false`.
+
+#### Related Templates
+
+signal_port
+
+</section>
+*/
+template signal_connect is connect {
+    interface signal {
+        saved bool high;
+    }
+
+    method set(conf_object_t *obj) default {
+        /* this is only called if the attribute actually changes */
+        default(obj);
+        if (signal.val && signal.high)
+            signal.signal_raise();
+    }
+
+    method raise() default {
+        if (!signal.high) {
+            signal.high = true;
+            if (signal.val)
+                signal.signal_raise();
```

Comment:
in some cases it can also be problematic to signal something upon connect, e.g. if connect is modified in flight for some hybrid substitution. But it's what the interface dictates, so I suppose we have no choice.